### PR TITLE
Engine bridge interface + Android implementation

### PR DESCRIPTION
**Pull request by sdsantos**
Closes https://github.com/ooni/probe-multiplatform/issues/17
Closes https://github.com/ooni/probe-multiplatform/issues/19

The `OonimkallBridge.kt` file is the most important part, the Bridge interface each platform has to implement. I believe I covered all our use cases, but please check.

I tried hiding the Session Context inside the bridge implementations. I feel it isn't too much on an hassle for the platforms. Not sure what timeout to use though.

@aanorbel Hope you can help with the iOS implementation :) (https://github.com/ooni/probe-multiplatform/issues/20)

### CODE REVIEW

**Comment by aanorbel** [comment]
Will look into that.

File: composeApp/src/commonMain/kotlin/org/ooni/probe/ui/dashboard/DashboardViewModel.kt [hunk 1]
```diff
@@ -73,7 +77,19 @@ class DashboardViewModel(
             TaskSettings(
                 name = "web_connectivity",
                 inputs = listOf("https://ooni.org"),
-                logLevel = "DEBUG2",
+                logLevel = TaskLogLevel.Info,
+                options =
+                    TaskSettings.Options(
+                        noCollector = true,
+                        softwareName = Config.BASE_SOFTWARE_NAME,
```
Comment (aanorbel) [inline composeApp/src/commonMain/kotlin/org/ooni/probe/ui/dashboard/DashboardViewModel.kt:84]: This `software_name` is set to `ooniprobe-android-unattended` when the tests are run from the background and `ooniprobe-android`.

See *Raw Measurement Data* in https://explorer.ooni.org/m/20240731235959.972250_IE_webconnectivity_50b43329c29ed566

File: composeApp/src/commonMain/kotlin/org/ooni/probe/ui/dashboard/DashboardViewModel.kt [hunk 2]
```diff
@@ -73,7 +77,19 @@ class DashboardViewModel(
             TaskSettings(
                 name = "web_connectivity",
```
Comment (aanorbel) [inline composeApp/src/commonMain/kotlin/org/ooni/probe/ui/dashboard/DashboardViewModel.kt:78]: We may want to start defining various test "classes" as some tests have different keys.

File: composeApp/src/commonMain/kotlin/org/ooni/engine/models/TaskSettings.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    @SerialName("version") val version: Int = 1,
+    @SerialName("log_level") val logLevel: TaskLogLevel,
+    @SerialName("disabled_events") val disabledEvents: List<String> = emptyList(),
+    @SerialName("proxy") val proxy: String? = null,
+    @SerialName("state_dir") val stateDir: String? = null,
+    @SerialName("temp_dir") val tempDir: String? = null,
+    @SerialName("tunnel_dir") val tunnelDir: String? = null,
+    @SerialName("assets_dir") val assetsDir: String? = null,
+    @SerialName("options") val options: Options,
+    @SerialName("annotations") val annotations: Annotations,
```
Comment (aanorbel) [inline composeApp/src/commonMain/kotlin/org/ooni/engine/models/TaskSettings.kt:19]: We should have a rule for property order. Having required properties first.

File: composeApp/src/commonMain/kotlin/org/ooni/engine/models/NetworkType.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+
+@Serializable(with = NetworkTypeSerializer::class)
+enum class NetworkType(val value: String) {
+    VPN("vpn"),
+    Wifi("wifi"),
+    Mobile("mobile"),
```
Comment (aanorbel) [inline composeApp/src/commonMain/kotlin/org/ooni/engine/models/NetworkType.kt:14]: We may want to start thinking of devices connected to the internet through `TRANSPORT_BLUETOOTH`, `TRANSPORT_ETHERNET`, `TRANSPORT_USB` especially on android that supports these options in the *Mobile Hotspot and Tethering* option.

File: composeApp/src/commonMain/kotlin/org/ooni/engine/models/TaskLogLevel.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+@Serializable(with = TaskLogLevelSerializer::class)
+enum class TaskLogLevel(val value: String) {
+    Debug("DEBUG2"),
+    Info("INFO"),
+}
+
+object TaskLogLevelSerializer : KSerializer<TaskLogLevel> {
+    override val descriptor =
+        PrimitiveSerialDescriptor("TaskLogLevel", PrimitiveKind.STRING)
```
Comment (aanorbel) [inline composeApp/src/commonMain/kotlin/org/ooni/engine/models/TaskLogLevel.kt:18]: Do we want to use class name here?

**Review by aanorbel — APPROVED** [review]
🐳

File: composeApp/src/commonMain/kotlin/org/ooni/probe/ui/dashboard/DashboardViewModel.kt (hunk 1 above)
Comment (sdsantos) [reply composeApp/src/commonMain/kotlin/org/ooni/probe/ui/dashboard/DashboardViewModel.kt:84]: For now, you can disregard the `DashboardViewModel.kt` code. I just kept the "Run Test" working to see if everything was fine. But we're never collecting/uploading, we still need to do the Engine SDK to properly set all configs.

File: composeApp/src/commonMain/kotlin/org/ooni/probe/ui/dashboard/DashboardViewModel.kt (hunk 2 above)
Comment (sdsantos) [reply composeApp/src/commonMain/kotlin/org/ooni/probe/ui/dashboard/DashboardViewModel.kt:78]: Same reply https://github.com/ooni/probe-multiplatform/pull/30#discussion_r1703870974

File: composeApp/src/commonMain/kotlin/org/ooni/engine/models/NetworkType.kt (hunk 4 above)
Comment (sdsantos) [reply composeApp/src/commonMain/kotlin/org/ooni/engine/models/NetworkType.kt:14]: Interesting. Didn't think about those, I was just following the current implementation on the Android app. Maybe that will be future work, after we reach feature parity?

File: composeApp/src/commonMain/kotlin/org/ooni/engine/models/TaskLogLevel.kt (hunk 5 above)
Comment (sdsantos) [reply composeApp/src/commonMain/kotlin/org/ooni/engine/models/TaskLogLevel.kt:18]: The official examples use the class name of the type we're serializing: https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/serializers.md#primitive-serializer

File: composeApp/src/commonMain/kotlin/org/ooni/engine/models/NetworkType.kt (hunk 4 above)
Comment (aanorbel) [reply composeApp/src/commonMain/kotlin/org/ooni/engine/models/NetworkType.kt:14]: yeah that works for me.

**Pull request merged**